contracts/ERC1238/extensions/ERC1238Holdable.sol

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "../ERC1238.sol";
+import "./IERC1238Holdable.sol";
+import "./IERC1238Holder.sol";
+
+/**
+ * @dev Proposal for ERC1238 tokens extension that allow addresses
+ * to hold tokens on behalf of others.
+ */
+abstract contract ERC1238Holdable is IERC1238Holdable, ERC1238 {
+    using Address for address;
+
+    // Mapping holder => id => balance
+    mapping(address => mapping(uint256 => uint256)) private _heldBalances;
+
+    function heldBalance(address holder, uint256 id) public view override returns (uint256) {
+        return _heldBalances[holder][id];
+    }
+
+    /**
+     * @dev Hooks into the minting flow to set the token recipient as first holder
+     * by default when tokens are minted.
+     */
+    function _beforeMint(
+        address,
+        address to,
+        uint256 id,
+        uint256 amount,
+        bytes memory
+    ) internal virtual override {
+        _heldBalances[to][id] += amount;
+    }
+
+    /**
+     * @dev Burns `amount` of tokens that are held by `holder` and owned by `from`.
+     * If `holder` is a smart contract and inherits {IERC1238Holder}, it notifies it to give it a chance to
+     * react to the burn and handle the operation how it sees fit.
+     *
+     * Requirements:
+     * - `holder` should hold at least the `amount` of tokens with the `id` passed
+     */
+    function _burnHeldTokens(
```

Comment:
I would still just call this a `burn` function. A burn may affect a token that is currently being held but it also may not. It should just attempt burn flows, check if its being held, if it is, then do more, if not, continue with burn flow.

Also I wouldn't use try catch for a specific error because the `onBurn` could halt the burn flow by reverting with a different error (or none altogether) where the point of using low level calls and checking failure was so that it doesn't stop burning. Assuming the holder both implements the expected interface AND will implement compatible logic doesn't work in the case where we don't want to delegate ANY responsibility to the holder contract.